packages/site/site-react/src/persistedQueries/persistedQueryRoute.ts

```diff
@@ -0,0 +1,115 @@
+    }
+
+    function inject(q: string): string {
+        // Find all fragment spreads in the query
+        const spreads = Array.from(q.matchAll(/\.\.\.(\w+)/g)).map((m) => m[1]);
+        let result = q;
+        for (const fragmentName of spreads) {
+            if (!injected.has(fragmentName) && fragmentsMap[fragmentName]) {
+                injected.add(fragmentName);
+                // Recursively inject fragments used by this fragment
+                result += `\n${inject(fragmentsMap[fragmentName])}`;
+            }
+        }
+        return result;
+    }
+    query = inject(query);
+
+    return query;
+}
+
+export async function persistedQueryRoute(
+    req: Request,
+    {
+        graphqlTarget,
+        headers: headersInit,
+        persistedQueriesPath = ".persisted-queries.json",
+    }: { graphqlTarget: string; headers?: HeadersInit; persistedQueriesPath: string },
+) {
+    if (!queryMap) {
+        await loadPersistedQueries(persistedQueriesPath);
+    }
+
+    let hash: string | null | undefined;
+    let variables: unknown;
+
+    if (req.method === "POST") {
+        const body = await req.json();
+        hash = body.extensions?.persistedQuery?.sha256Hash;
+        variables = body.variables;
+    } else if (req.method === "GET") {
+        const url = new URL(req.url);
+        hash = url.searchParams.get("extensions.persistedQuery.sha256Hash");
+        const variablesParam = url.searchParams.get("variables");
+        variables = variablesParam ? JSON.parse(variablesParam) : undefined;
+    } else {
+        return Response.json({ error: "MethodNotAllowed" }, { status: 405 });
+    }
+
+    // Only allow persisted queries
+    if (!hash) {
+        return Response.json({ error: "OnlyPersistedQueriesAllowed" }, { status: 400 });
+    }
+
+    let query = queryMap[hash];
+    if (!query) {
+        return Response.json({ error: "PersistedQueryNotFound", hash }, { status: 400 });
+    }
+    query = injectFragments(query);
+
+    // CSRF protection for GET requests, similar to Apollo Server's CSRF protection
+    if (req.method === "GET") {
+        const apolloPreflight = req.headers.get("Apollo-Require-Preflight");
+        if (apolloPreflight !== "true") {
+            return Response.json({ error: "CSRFProtection", message: "Missing Apollo-Require-Preflight header" }, { status: 403 });
+        }
+    }
+
+    const headers = new Headers(headersInit);
+    headers.set("Content-Type", "application/json");
+
+    // Forward to actual GraphQL server
+    const upstreamRes = await fetch(graphqlTarget, {
+        method: "POST",
+        headers,
+        body: JSON.stringify({ query: query, variables }),
+    });
+
+    const responseHeaders: Record<string, string> = {
+        "Content-Type": "application/json",
+    };
+    if (req.method === "GET") {
+        // Cache for 7.5 minutes (450 seconds) in CDNs and browsers
+        responseHeaders["Cache-Control"] = "public, max-age=450";
+    }
```

Comment:
Should we support overriding headers?

Reply:
yes, hardcoding 7.5 mintues is not good. what api could we add here?

Reply:
> Might be enough, if null (or 0?) means disable.

Yes, any negative value or 0 should disable it. I've seen this behavior in some libraries.